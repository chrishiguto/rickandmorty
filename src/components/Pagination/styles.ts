import styled, { css, DefaultTheme } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.nav`
  display: flex;
  align-items: center;
`

export const Pages = styled.ul`
  display: inline-flex;
  align-items: center;
  gap: 2.4rem;
  margin-right: 4.8rem;
  margin-left: 4.8rem;
  list-style: none;
`

const Button = styled.button`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;

    outline: none;
    border: none;
    background: none;

    font-size: ${theme.font.sizes.large};
    color: ${theme.colors.lightGray};
    line-height: 1;
    height: 3.2rem;
    width: 3.2rem;
    transition: ${theme.transition.fast};
    border-radius: 50%;

    &:not(:disabled):hover {
      background: ${theme.colors.gray};
    }
  `}
`

const pageButtonModifiers = {
  active: (theme: DefaultTheme) => css`
    color: ${theme.colors.yellow};
  `
}

export const PageButton = styled(Button)``

type PageProps = {
  isActive?: boolean
}

export const Page = styled.li<PageProps>`
  ${({ theme, isActive }) => css`
    display: ${isActive ? 'block' : 'none'};

    ${media.greaterThan('medium')`
      display: block;
    `}

    ${PageButton} {
      ${!!isActive && pageButtonModifiers.active(theme)};
    }
  `}
`

export const PageInterval = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: center;

    font-size: ${theme.font.sizes.large};
    color: ${theme.colors.lightGray};
    line-height: 1;
    height: 3.2rem;
    width: 3.2rem;
    border-radius: 50%;
  `}
`

const arrowButtonModifiers = {
  disabled: (theme: DefaultTheme) => css`
    cursor: not-allowed;

    > svg {
      color: ${theme.colors.gray};
    }
  `
}

export const ArrowButton = styled(Button)`
  ${({ theme, disabled }) => css`
    > svg {
      color: ${theme.colors.white};
    }

    ${disabled && arrowButtonModifiers.disabled(theme)}
  `}
`
