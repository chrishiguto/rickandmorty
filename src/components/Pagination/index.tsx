import React, { useCallback } from 'react'
import { ChevronLeft, ChevronRight } from '@styled-icons/feather'

import * as S from './styles'

export type PaginationProps = {
  prev: number | null
  next: number | null
  pages: number
  activePage: number
  onChange: (value: number) => void
}

type PageProps = {
  page: number
  isActive?: boolean
  onClick: () => void
}

const Page = ({ isActive, page, onClick }: PageProps) => (
  <S.Page isActive={!!isActive}>
    <S.PageButton onClick={onClick}>{page}</S.PageButton>
  </S.Page>
)

const More = () => (
  <S.Page>
    <S.PageInterval>...</S.PageInterval>
  </S.Page>
)

const Pagination = ({
  prev,
  next,
  pages,
  activePage,
  onChange
}: PaginationProps) => {
  const range = useCallback(
    (min: number, max: number) =>
      Array.from({ length: max - min + 1 }, (_, i) => min + i),
    []
  )

  const startPage = Math.max(2, activePage - 1)
  const endPage = Math.min(pages - 1, activePage + 1)
  const pagesToRender = range(startPage, endPage)

  const hasLeftPages = () => startPage > 2
  const hasRightPages = () => pages - endPage > 1

  const handleFirstPage = () => onChange(1)
  const handleLastPage = () => onChange(pages)

  const renderPagination = (
    hasLeftPages: boolean,
    hasRightPages: boolean,
    pagesToRender: number[],
    pages: number
  ) => (
    <>
      {hasLeftPages && !hasRightPages && (
        <>
          <Page
            page={1}
            isActive={1 === activePage}
            onClick={handleFirstPage}
          />
          <More />
          {pagesToRender.map((page) => (
            <Page
              key={page}
              page={page}
              isActive={page === activePage}
              onClick={() => onChange(page)}
            />
          ))}
          <Page
            page={pages}
            isActive={pages === activePage}
            onClick={handleLastPage}
          />
        </>
      )}
      {!hasLeftPages && hasRightPages && (
        <>
          <Page
            page={1}
            isActive={1 === activePage}
            onClick={handleFirstPage}
          />
          {pagesToRender.map((page) => (
            <Page
              key={page}
              page={page}
              isActive={page === activePage}
              onClick={() => onChange(page)}
            />
          ))}
          <More />
          <Page
            page={pages}
            isActive={pages === activePage}
            onClick={handleLastPage}
          />
        </>
      )}
      {hasLeftPages && hasRightPages && (
        <>
          <Page
            page={1}
            isActive={1 === activePage}
            onClick={handleFirstPage}
          />
          <More />
          {pagesToRender.map((page) => (
            <Page
              key={page}
              page={page}
              isActive={page === activePage}
              onClick={() => onChange(page)}
            />
          ))}
          <More />
          <Page
            page={pages}
            isActive={pages === activePage}
            onClick={handleLastPage}
          />
        </>
      )}
      {!hasLeftPages && !hasRightPages && (
        <Page page={1} isActive={1 === activePage} onClick={handleFirstPage} />
      )}
    </>
  )

  return (
    <S.Wrapper>
      <S.ArrowButton
        aria-label="Previous page button"
        onClick={prev === null ? undefined : () => onChange(prev)}
        disabled={prev === null}
      >
        <ChevronLeft size={24} strokeWidth={2} />
      </S.ArrowButton>
      <S.Pages>
        {renderPagination(
          hasLeftPages(),
          hasRightPages(),
          pagesToRender,
          pages
        )}
      </S.Pages>
      <S.ArrowButton
        aria-label="Next page button"
        onClick={next === null ? undefined : () => onChange(next)}
        disabled={next === null}
      >
        <ChevronRight size={24} strokeWidth={2} />
      </S.ArrowButton>
    </S.Wrapper>
  )
}

export default Pagination
