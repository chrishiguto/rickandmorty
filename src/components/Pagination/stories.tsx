import { Story, Meta } from '@storybook/react/types-6-0'
import Pagination, { PaginationProps } from '.'

export default {
  title: 'Pagination',
  component: Pagination,
  parameters: {
    backgrounds: {
      default: 'dark'
    }
  },
  argTypes: {
    onChange: { action: 'page clicked' }
  }
} as Meta

export const Default: Story<PaginationProps> = (args) => (
  <Pagination {...args} />
)

Default.args = {
  prev: 4,
  next: 6,
  pages: 10,
  activePage: 5
}
