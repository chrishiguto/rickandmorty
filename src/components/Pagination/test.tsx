import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'

import Pagination from '.'

describe('<Pagination />', () => {
  it('should render the pagination with left and right pages and trigger change function if clicked', async () => {
    const mockFn = jest.fn()

    const { container } = renderWithTheme(
      <Pagination
        pages={10}
        activePage={5}
        next={6}
        prev={4}
        onChange={mockFn}
      />
    )

    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getAllByText('...')).toHaveLength(2)
    expect(screen.getByText('4')).toBeInTheDocument()
    expect(screen.getByText('5')).toBeInTheDocument()
    expect(screen.getByText('6')).toBeInTheDocument()
    expect(screen.getByText('10')).toBeInTheDocument()

    userEvent.click(screen.getByText('4'))
    userEvent.click(screen.getByLabelText(/previous page button/i))
    userEvent.click(screen.getByLabelText(/next page button/i))

    await waitFor(() => {
      expect(mockFn).toBeCalledTimes(3)
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pagination with left pages and trigger change function if clicked', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={10}
        activePage={9}
        next={10}
        prev={8}
        onChange={mockFn}
      />
    )

    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('...')).toBeInTheDocument()
    expect(screen.getByText('8')).toBeInTheDocument()
    expect(screen.getByText('9')).toBeInTheDocument()
    expect(screen.getByText('10')).toBeInTheDocument()

    userEvent.click(screen.getByText('8'))

    await waitFor(() => {
      expect(mockFn).toBeCalledTimes(1)
    })
  })

  it('should render the pagination with right pages and trigger change function if clicked', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={10}
        activePage={2}
        next={3}
        prev={1}
        onChange={mockFn}
      />
    )

    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
    expect(screen.getByText('3')).toBeInTheDocument()
    expect(screen.getByText('...')).toBeInTheDocument()
    expect(screen.getByText('10')).toBeInTheDocument()

    userEvent.click(screen.getByText('3'))

    await waitFor(() => {
      expect(mockFn).toBeCalledTimes(1)
    })
  })

  it('should render the active page', () => {
    renderWithTheme(
      <Pagination
        pages={1}
        activePage={1}
        next={null}
        prev={null}
        onChange={jest.fn}
      />
    )

    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.queryByText('...')).not.toBeInTheDocument()
  })

  it('should render the pagination with left arrow disabled', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={2}
        activePage={1}
        next={2}
        prev={null}
        onChange={mockFn}
      />
    )

    const button = screen.getByLabelText(/previous page button/i)

    expect(button).toBeDisabled()

    userEvent.click(button)

    await waitFor(() => {
      expect(mockFn).not.toBeCalled()
    })
  })

  it('should render the pagination with right arrow disabled', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={2}
        activePage={2}
        next={null}
        prev={1}
        onChange={mockFn}
      />
    )

    const button = screen.getByLabelText(/next page button/i)

    expect(button).toBeDisabled()

    userEvent.click(button)

    await waitFor(() => {
      expect(mockFn).not.toBeCalled()
    })
  })

  it('should go to first page if clicked', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={10}
        activePage={9}
        next={10}
        prev={8}
        onChange={mockFn}
      />
    )

    const button = screen.getByText('1')

    userEvent.click(button)

    await waitFor(() => expect(mockFn).toBeCalledTimes(1))
  })

  it('should go to last page if clicked', async () => {
    const mockFn = jest.fn()

    renderWithTheme(
      <Pagination
        pages={10}
        activePage={5}
        next={6}
        prev={4}
        onChange={mockFn}
      />
    )

    const button = screen.getByText('10')

    userEvent.click(button)

    await waitFor(() => expect(mockFn).toBeCalledTimes(1))
  })
})
