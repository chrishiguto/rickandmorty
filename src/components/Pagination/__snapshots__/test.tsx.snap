// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Pagination /> should render the pagination with left and right pages and trigger change function if clicked 1`
] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2.4rem;
  margin-right: 4.8rem;
  margin-left: 4.8rem;
  list-style: none;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  border: none;
  background: none;
  font-size: 1.8rem;
  color: #D3D3D3;
  line-height: 1;
  height: 3.2rem;
  width: 3.2rem;
  -webkit-transition: 0.1s cubic-bezier(0.4,0,0.2,1);
  transition: 0.1s cubic-bezier(0.4,0,0.2,1);
  border-radius: 50%;
}

.c1:not(:disabled):hover {
  background: #606060;
}

.c5 {
  display: none;
}

.c7 {
  display: block;
}

.c7 .{
  color: #cbd736;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 1.8rem;
  color: #D3D3D3;
  line-height: 1;
  height: 3.2rem;
  width: 3.2rem;
  border-radius: 50%;
}

.c2 > svg {
  color: #ffffff;
}

@media (min-width:768px) {
  .c5 {
    display: block;
  }
}

@media (min-width:768px) {
  .c7 {
    display: block;
  }
}

<nav
  class="c0"
>
  <button
    aria-label="Previous page button"
    class="c1 c2"
  >
    <svg
      aria-hidden="true"
      class="c3"
      fill="none"
      focusable="false"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="15 18 9 12 15 6"
      />
    </svg>
  </button>
  <ul
    class="c4"
  >
    <li
      class="c5"
    >
      <button
        class="c1"
      >
        1
      </button>
    </li>
    <li
      class="c5"
    >
      <div
        class="c6"
      >
        ...
      </div>
    </li>
    <li
      class="c5"
    >
      <button
        class="c1"
      >
        4
      </button>
    </li>
    <li
      class="c7"
    >
      <button
        class="c1"
      >
        5
      </button>
    </li>
    <li
      class="c5"
    >
      <button
        class="c1"
      >
        6
      </button>
    </li>
    <li
      class="c5"
    >
      <div
        class="c6"
      >
        ...
      </div>
    </li>
    <li
      class="c5"
    >
      <button
        class="c1"
      >
        10
      </button>
    </li>
  </ul>
  <button
    aria-label="Next page button"
    class="c1 c2"
  >
    <svg
      aria-hidden="true"
      class="c3"
      fill="none"
      focusable="false"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </button>
</nav>
`
